src/clj_puppetdb/http.clj

```diff
@@ -33,31 +32,25 @@
   (let [opts (assoc opts :as :stream)
         info (select-keys opts connection-relevant-opts)
         info (assoc info :host host)]
-    (fn
-      ; This arity-overloaded variant of the function is the one called by the `GET` function.
-      ; The `this` parameter refers under normal circumstances to this very function and it is
-      ; needed to be able to call another arity-overloaded variant of the function from the
-      ; function body. When the `this` parameter refers to a different function then it
-      ; effectively overrides the other arity-overloaded variants of the function.
-      ([this path params]
-       (let [query (if (empty? params)
-                     (str host path)
-                     (str host path \? (-> params
-                                           q/params->json
-                                           map->query)))]
-         ; now call the single argument variant of the function to execute the query
-         (this query)))
-
-      ; This arity-overloaded variant of the function is the one which actually executes
-      ; the PDB query.
-      ([query]
-       (log/debug (str "GET:" query))
-       (http/get query opts))
-
-      ; This arity-overloaded variant of the function returns a map with information about
-      ; this client.
-      ([]
-       info))))
+    (reify
+      PdbClient
+      (pdb-get [this path params]
+        (pdb-get this this path params))
+      ; this arity is meant to support a kind of polymorphism
+      (pdb-get [_ that path params]
```

Comment:
Note also that while the first argument of the ```(pdb-get ...)``` is not used in the function body, it is used by the clojure to dispatch the call the the appropriate implementation of the function based on the concrete type of the first parameter. (My assumption is that this is implemented by taking advantage by the standard java virtual method call mechanism.)